File: apps/ios/ios-rust-test/AppContainer.swift
import Foundation

/// Cableado manual, sin librería de DI.
///
/// Con cinco pantallas y tres dependencias, explícito gana: se lee de arriba abajo y el
/// compilador lo verifica entero. Un contenedor de DI no sería *más* desacoplado, sería
/// más automático — y fallaría más tarde.
struct AppContainer {
    let core: CoreFinanciero
    let contract: ContractSource
    let messages: ContractMessages

    /// Lanza si el contrato no está en el bundle: es un fallo de build, no de dominio, y
    /// tiene que verse al arrancar en vez de dejar la pantalla de Transferencia sin
    /// cuentas y sin explicación.
    init(bundle: Bundle = .main) throws {
        core = UniffiCoreFinanciero()
        contract = try BundleContractSource(bundle: bundle)
        messages = ContractMessages(source: try BundleMessageSource(bundle: bundle))
    }
}

File: apps/android/app/src/main/java/dev/tohure/android_rust_test/AppContainer.kt
package dev.tohure.android_rust_test

import android.content.Context
import dev.tohure.android_rust_test.adapter.ContractMessages
import dev.tohure.android_rust_test.adapter.CoreFinanciero
import dev.tohure.android_rust_test.adapter.UniffiCoreFinanciero
import dev.tohure.android_rust_test.contract.AssetContractSource
import dev.tohure.android_rust_test.contract.AssetMessageSource
import dev.tohure.android_rust_test.contract.ContractSource

/**
 * Cableado manual, sin librería de DI.
 *
 * Hilt trae KSP y tiempo de build; Koin trae resolución en runtime que falla tarde.
 * Ninguno de los dos es *más* desacoplado que esto: son más automáticos. Con cinco
 * pantallas y tres dependencias, explícito gana — se lee de arriba abajo y el compilador
 * lo verifica entero.
 */
class AppContainer(context: Context) {
    private val assets = context.applicationContext.assets

    val core: CoreFinanciero = UniffiCoreFinanciero()
    val contract: ContractSource = AssetContractSource(assets)
    val messages: ContractMessages = ContractMessages(AssetMessageSource(assets))
}
